File: app/src/test/java/com/example/david/lists/view/itemlist/ItemListLogicTest.java
package com.example.david.lists.view.itemlist;

import com.example.david.lists.SchedulerProviderMockInit;
import com.example.david.lists.data.datamodel.Item;
import com.example.david.lists.data.repository.IRepositoryContract.Repository;
import com.example.david.lists.util.ISchedulerProviderContract;
import com.example.david.lists.view.itemlist.IItemViewContract.Adapter;
import com.example.david.lists.view.itemlist.IItemViewContract.View;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.disposables.CompositeDisposable;

import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class ItemListLogicTest {

    @Mock
    private View view;

    @Mock
    private IItemViewContract.ViewModel viewModel;

    @Mock
    private Repository repo;

    @Mock
    private ISchedulerProviderContract schedulerProvider;

    @Spy
    private CompositeDisposable disposable;


    @InjectMocks
    private ItemListLogic logic;


    @Mock
    private Adapter adapter;

    private String userListId = "qwerty";

    private String id = "id_one";
    private String title = "title_one";
    private int position = 0;
    private Item item = new Item(id, new Item(title, position, userListId));

    private String id_two = "id_two";
    private String title_two = "title_two";
    private int position_two = 1;
    private Item item_two = new Item(id_two, new Item(title_two, position_two, userListId));

    private List<Item> itemList = new ArrayList<>(Collections.singletonList(item));

    private int invalidPosition = -1;
    private String errorMsg = "error";


    @Before
    public void setUp() {
        SchedulerProviderMockInit.init(schedulerProvider);
    }


    /**
     * Normal behavior
     * - Set View state loading.
     * - Observe deleted UserLists Flowable from repo.
     * - Get List from repo.
     * - Save List to ViewModel.
     * - Submit the List to the View.
     * - Set View state display list.
     */
    @Test
    public void onStart() {
        when(viewModel.getUserListId()).thenReturn(userListId);
        when(viewModel.getViewData()).thenReturn(itemList);
        when(repo.getEventUserListDeleted()).thenReturn(Flowable.just(new ArrayList<>()));
        when(repo.getItems(userListId)).thenReturn(Flowable.just(itemList));

        logic.onStart();

        verify(view).setStateLoading();
        verify(viewModel).setViewData(itemList);
        verify(view).submitList(itemList);
        verify(view).setStateDisplayList();
    }

    /**
     * Error behavior - empty List from repo,
     * - Set View state loading.
     * - Observe deleted UserLists Flowable from repo.
     * - Get List from repo.
     * - Save List to ViewModel.
     * - Submit the List to the View.
     * - Set View state error with message from the ViewModel.
     */
    @Test
    public void onStartEmptyList() {
        List<Item> emptyList = new ArrayList<>();

        when(viewModel.getUserListId()).thenReturn(userListId);
        when(viewModel.getViewData()).thenReturn(emptyList);
        when(viewModel.getErrorMsgEmptyList()).thenReturn(errorMsg);
        when(repo.getEventUserListDeleted()).thenReturn(Flowable.just(new ArrayList<>()));
        when(repo.getItems(userListId)).thenReturn(Flowable.just(emptyList));

        logic.onStart();

        verify(view).setStateLoading();
        verify(viewModel).setViewData(emptyList);
        verify(view).submitList(emptyList);
        verify(view).setStateError(errorMsg);
    }

    /**
     * Error behavior - repo throws an error,
     * - Set View state loading.
     * - Observe deleted UserLists Flowable from repo.
     * - Get List from repo.
     * - Repo throws an error.
     * - Set View state error with message from the ViewModel.
     */
    @Test
    public void onStartRepoThrowsError() {
        Throwable throwable = new Throwable();

        when(viewModel.getUserListId()).thenReturn(userListId);
        when(viewModel.getErrorMsg()).thenReturn(errorMsg);
        when(repo.getEventUserListDeleted()).thenReturn(Flowable.just(new ArrayList<>()));
        when(repo.getItems(userListId)).thenReturn(Flowable.error(throwable));

        logic.onStart();

        verify(view).setStateLoading();
        verify(view).setStateError(errorMsg);
    }


    /**
     * Normal behavior
     * - Invoke {@link View#openAddDialog(String, int)} with the
     * UserListId and the view data's List size.
     */
    @Test
    public void add() {
        when(viewModel.getUserListId()).thenReturn(userListId);
        when(viewModel.getViewData()).thenReturn(itemList);

        logic.add();

        verify(view).openAddDialog(userListId, itemList.size());
    }


    /**
     * Normal behavior
     * - Get the Item at the passed-in position.
     * - Invoke {@link View#openEditDialog(Item)}.
     */
    @Test
    public void edit() {
        int editPosition = 0;
        List<Item> editItems = new ArrayList<>();
        editItems.add(item);

        when(viewModel.getViewData()).thenReturn(editItems);

        logic.edit(editPosition);

        verify(view).openEditDialog(item);
    }

    /**
     * Error behavior - invalid position,
     * - Throw an Exception.
     */
    @Test(expected = IllegalArgumentException.class)
    public void editInvalidPosition() {
        logic.edit(invalidPosition);
    }


    /**
     * Normal behavior
     * - Update the Adapter with {@link Adapter#move(int, int)}.
     * - Update the view data.
     */
    @Test
    public void dragging() {
        int fromPos = 0;
        int toPos = 1;

        List<Item> draggingItems = new ArrayList<>();
        draggingItems.add(item);
        draggingItems.add(item_two);

        when(viewModel.getViewData()).thenReturn(draggingItems);

        logic.dragging(fromPos, toPos, adapter);

        verify(adapter).move(fromPos, toPos);
        assertThat(draggingItems, is(contains(item_two, item)));
    }


    /**
     * Normal behavior
     * - Gets Item at passed-in position.
     * - Invokes {@link Repository#updateItemPosition(Item, int, int)}
     */
    @Test
    public void movedPermanently() {
        int newPosition = 0;
        List<Item> moveItems = new ArrayList<>();
        moveItems.add(item);

        when(viewModel.getViewData()).thenReturn(moveItems);

        logic.movedPermanently(newPosition);

        verify(repo).updateItemPosition(item, item.getPosition(), newPosition);
    }

    /**
     * Error behavior - invalid position,
     * - Throws an Exception
     */
    @Test(expected = IllegalArgumentException.class)
    public void movedPermanentlyInvalidPosition() {
        logic.movedPermanently(invalidPosition);
    }


    /**
     * Normal behavior,
     * - Remove from the Adapter.
     * - Save the deleted UserList itself to the ViewModel's temp list.
     * - Save the deleted UserList's position to the ViewModel.
     * - Remove from the view data in the ViewModel.
     * - Notify user of deletion.
     */
    @Test
    public void delete() {
        int deletedPosition = 0;
        List<Item> deletedItems = new ArrayList<>();
        deletedItems.add(item);
        List<Item> tempList = new ArrayList<>();

        when(viewModel.getViewData()).thenReturn(deletedItems);
        when(viewModel.getTempList()).thenReturn(tempList);
        when(viewModel.getMsgItemDeleted()).thenReturn(errorMsg);

        logic.delete(deletedPosition, adapter);

        verify(adapter).remove(position);
        verify(viewModel).setTempPosition(deletedPosition);
        verify(view).notifyUserOfDeletion(errorMsg);
        assertThat(deletedItems.isEmpty(), is(true));
        assertThat(tempList, is(contains(item)));
    }

    /**
     * Error behavior - invalid position,
     * - Throws an exception.
     */
    @Test(expected = IllegalArgumentException.class)
    public void deleteInvalidPosition() {
        logic.delete(invalidPosition, adapter);
    }


    /**
     * Normal behavior
     * - Re-add to the adapter.
     * - Re-add to the view data List.
     * - Remove the last Item from the temp List.
     * - Because the temp List is not empty, pass the temp List to the repo for deletion.
     * - Clear the temp List.
     */
    @Test
    public void undoRecentDeletion() {
        int undoPosition = 0;
        List<Item> tempList = new ArrayList<>();
        tempList.add(item);
        tempList.add(item_two);
        List<Item> viewDataList = new ArrayList<>();

        when(viewModel.getTempList()).thenReturn(tempList);
        when(viewModel.getTempPosition()).thenReturn(undoPosition);
        when(viewModel.getViewData()).thenReturn(viewDataList);

        logic.undoRecentDeletion(adapter);

        verify(adapter).reAdd(undoPosition, item_two);
        verify(repo).deleteItems(tempList);
        assertThat(viewDataList, is(contains(item_two)));
        assertThat(tempList.isEmpty(), is(true));
    }

    /**
     * Error behavior - empty temp List,
     * - Throws an Exception.
     */
    @Test(expected = UnsupportedOperationException.class)
    public void undoRecentDeletionEmptyTempList() {
        when(viewModel.getTempList()).thenReturn(new ArrayList<>());

        logic.undoRecentDeletion(adapter);
    }

    /**
     * Error behavior - invalid temp position,
     * - Throws an Exception.
     */
    @Test(expected = UnsupportedOperationException.class)
    public void undoRecentDeletionInvalidTempPosition() {
        when(viewModel.getTempList()).thenReturn(itemList);
        when(viewModel.getTempPosition()).thenReturn(invalidPosition);

        logic.undoRecentDeletion(adapter);
    }


    /**
     * Normal behavior
     * - Check if the temp List is empty.
     * - Pass the temp List {@link Repository#deleteItems(List)}.
     * - Clear the temp List.
     */
    @Test
    public void deletionNotificationTimedOut() {
        when(viewModel.getTempList()).thenReturn(itemList);

        logic.deletionNotificationTimedOut();

        verify(repo).deleteItems(itemList);
        assertThat(itemList.isEmpty(), is(true));
    }

    /**
     * Error behavior - empty temp List,
     * - Method returns without invoking the repo.
     */
    @Test
    public void deletionNotificationTimedOutEmptyTempList() {
        when(viewModel.getTempList()).thenReturn(new ArrayList<>());

        logic.deletionNotificationTimedOut();

        verify(repo, never()).deleteItems(anyList());
    }


    /**
     * Normal behavior
     * - Verify {@link CompositeDisposable#clear()} is invoked.
     */
    @Test
    public void onDestroy() {
        logic.onDestroy();
        verify(disposable).clear();
    }
}

File: app/src/test/java/com/example/david/lists/view/itemlist/ItemListLogicTest.kt
package com.example.david.lists.view.itemlist

import com.example.david.lists.SchedulerProviderMockInit
import com.example.david.lists.util.ISchedulerProviderContract
import com.example.domain.datamodel.Item
import com.example.domain.datamodel.UserList
import com.example.domain.repository.IRepositoryContract
import io.mockk.*
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource

class ItemListLogicTest {

    private val view = mockk<IItemViewContract.View>(relaxUnitFun = true)

    private val viewModel = mockk<IItemViewContract.ViewModel>(relaxUnitFun = true)

    private val repo = mockk<IRepositoryContract.Repository>(relaxUnitFun = true)

    private val schedulerProvider = mockk<ISchedulerProviderContract>()

    private val disposable = spyk<CompositeDisposable>()


    private val logic = ItemListLogic(view, viewModel, repo, schedulerProvider, disposable)


    private val adapter = mockk<IItemViewContract.Adapter>(relaxUnitFun = true)


    private val id = "id"
    private val title = "title"
    private val position = 0
    private val userListId = "qwerty"
    private val itemOne = Item(title, position, userListId, id)
    private val itemTwo = Item(title + 2, position + 1, userListId, id + 2)

    private val message = "message"

    private val emptyList = mutableListOf<Item>()


    @BeforeEach
    fun init() {
        clearAllMocks()

        SchedulerProviderMockInit.init(schedulerProvider)

        every { viewModel.userListId } returns userListId
        every { repo.userListDeletedObservable } returns Flowable.just(emptyList())
    }


    @Nested
    inner class OnStart {
        /**
         * - Set the View's data with data from the ViewModel - because the ViewModel's view data is not empty.
         * - Get List from repo.
         * - Save List to ViewModel.
         * - Set the View's data
         * - Set View state display list.
         */
        @Test
        fun `onStart - ViewModel has View Data`() {
            val itemList = mutableListOf(itemOne)

            every { repo.getItems(userListId) } returns Flowable.just(itemList)
            every { viewModel.viewData } returns itemList

            logic.onStart()

            verify(exactly = 2) { view.setViewData(itemList) }
            verify { viewModel.viewData = itemList }
            verify { view.setStateDisplayList() }
        }

        /**
         * - Set View state loading.
         *   - Because the ViewModel's ViewData is empty.
         * - Get List from repo.
         * - Save List to ViewModel.
         * - Submit the List to the View.
         * - Set View state error with message from the ViewModel.
         */
        @Test
        fun `onStart - Empty List from repo`() {
            every { repo.getItems(userListId) } returns Flowable.just(emptyList)
            every { viewModel.viewData } returns emptyList
            every { viewModel.errorMsgEmptyList } returns message

            logic.onStart()

            verify { view.setStateLoading() }
            verify { viewModel.viewData = emptyList }
            verify { view.setViewData(emptyList) }
            verify { view.setStateError(message) }
        }

        /**
         * - Set View state loading.
         *   - Because the ViewModel's ViewData is empty.
         * - Get List from repo.
         * - Repo throws an error.
         * - Set View state error with message from ViewModel.
         */
        @Test
        fun `onStart - Repo throws an error`() {
            val throwable = Throwable()

            every { viewModel.viewData } returns emptyList
            every { repo.getItems(userListId) } returns Flowable.error(throwable)
            every { viewModel.errorMsg } returns message

            logic.onStart()

            verify { view.setStateLoading() }
            verify { view.setStateError(message) }
        }
    }


    @Nested
    inner class ObserveDeletedUserLists {
        /**
         * - Observe the observer from the Repo.
         * - Check each UserList's ID against the UserListId in the ViewModel.
         * - If they are equal - they will in this test - display a message and finish the View.
         */
        @Test
        fun `observeDeletedUserLists - UserListId matches`() {
            val userListList = listOf(UserList(title, position, userListId))
            val itemList = mutableListOf(itemOne)

            every { viewModel.viewData } returns itemList
            every { repo.getItems(userListId) } returns Flowable.just(itemList)
            every { repo.userListDeletedObservable } returns Flowable.just(userListList)
            every { viewModel.getMsgListDeleted(title) } returns message

            logic.onStart()

            verify { view.showMessage(message) }
            verify { view.finishView() }
        }

        /**
         * - Observe the observer from the Repo.
         * - Check each UserList's ID against the UserListId in the ViewModel.
         * - In this test they will not be equal, thus nothing will happen.
         */
        @Test
        fun `observeDeletedUserLists - UserListId does not match`() {
            val userListIdShouldNotMatch = ""
            val userListList = listOf(UserList(title, position, userListIdShouldNotMatch))
            val itemList = mutableListOf(itemOne)

            every { viewModel.viewData } returns itemList
            every { repo.getItems(any()) } returns Flowable.just(itemList)
            every { repo.userListDeletedObservable } returns Flowable.just(userListList)
            every { viewModel.getMsgListDeleted(title) } returns message

            logic.onStart()

            verify(exactly = 0) { view.showMessage(message) }
            verify(exactly = 0) { view.finishView() }
        }
    }


    /**
     * - [IItemViewContract.View.openAddDialog] is invoked with the current size.
     * of List containing the view data.
     */
    @Test
    fun add() {
        val itemList = mutableListOf(itemOne)
        val size = itemList.size

        every { viewModel.viewData } returns itemList

        logic.add()

        verify { view.openAddDialog(userListId, size) }
    }


    @Nested
    inner class Edit {
        /**
         * - Get the Item at the passed-in position.
         * - Pass that Item to [IItemViewContract.View.openEditDialog].
         */
        @Test
        fun edit() {
            val position = 0
            val itemList = mutableListOf(itemOne)

            every { viewModel.viewData } returns itemList

            logic.edit(position)

            verify { view.openEditDialog(itemOne) }
        }

        /**
         * - Attempt to get the selected UserList from the ViewModel with the invalid position.
         * - Exception is thrown.
         */
        @Test
        fun `edit - Invalid Position`() {
            val invalidPosition = -1
            val itemList = mutableListOf(itemOne)

            every { viewModel.viewData } returns itemList

            assertThrows<IndexOutOfBoundsException> {
                logic.edit(invalidPosition)
            }
        }
    }


    @Nested
    inner class Dragging {
        /**
         * - Updated the Adapter with [IItemViewContract.Adapter.move].
         * - Update the ViewModel's view data - verify that the order has changed.
         */
        @Test
        fun dragging() {
            val fromPosition = 0
            val toPosition = 1
            val itemList = mutableListOf(itemOne, itemTwo)

            every { viewModel.viewData } returns itemList

            logic.dragging(fromPosition, toPosition, adapter)

            verify { adapter.move(fromPosition, toPosition) }
            // Assert that the ViewData was updated.
            assertThat(itemList).containsExactly(itemTwo, itemOne)
        }

        /**
         * - Attempt to re-arrange the ViewData with the invalid positions.
         * - Exception is thrown.
         */
        @ParameterizedTest
        @CsvSource("-1, 0", "0, -1", "-1, -1")
        fun `dragging - Invalid positions`(fromPosition: Int, toPosition: Int) {
            val itemList = mutableListOf(itemOne, itemTwo)

            every { viewModel.viewData } returns itemList

            assertThrows<IndexOutOfBoundsException> {
                logic.dragging(fromPosition, toPosition, adapter)
            }
        }
    }


    @Nested
    inner class MovePermanently {
        /**
         * - Get the moved Item from the ViewModel.
         * - Invoke [IRepositoryContract.Repository.updateItemPosition].
         */
        @Test
        fun movedPermanently() {
            val newPosition = 0
            val itemList = mutableListOf(itemOne)

            every { viewModel.viewData } returns itemList

            logic.movedPermanently(newPosition)

            verify { repo.updateItemPosition(itemOne, itemOne.position, newPosition) }
        }

        /**
         * - Attempt to get the selected Item from the ViewModel with the invalid position.
         * - Exception is thrown.
         */
        @Test
        fun `movedPermanently - InvalidPosition`() {
            val invalidPosition = -1
            val itemList = mutableListOf(itemOne)

            every { viewModel.viewData } returns itemList

            assertThrows<IndexOutOfBoundsException> {
                logic.movedPermanently(invalidPosition)
            }
        }
    }


    @Nested
    inner class Delete {
        /**
         * - Get the deleted Item from the ViewModel.
         * - Remove from the Adapter.
         * - Save the deleted Item itself to the ViewModel's temp list.
         * - Save the deleted Item's position to the ViewModel.
         * - Remove from the view data in the ViewModel.
         * - Notify user of deletion.
         */
        @Test
        fun delete() {
            val position = 0
            val itemList = mutableListOf(itemOne)
            val tempItemLists = mutableListOf<Item>()

            every { viewModel.viewData } returns itemList
            every { viewModel.tempList } returns tempItemLists
            every { viewModel.msgDeletion } returns message

            logic.delete(position, adapter)

            verify { adapter.remove(position) }
            assertThat(tempItemLists).containsExactly(itemOne)
            verify { viewModel.tempPosition = position }
            assertThat(itemList.isEmpty()).isTrue()
            verify { view.notifyUserOfDeletion(message) }
        }

        /**
         * - Attempt to get the selected Item from the ViewModel with the invalid position.
         * - Exception is thrown
         */
        @Test
        fun `delete - Invalid Position`() {
            val invalidPosition = -1
            val itemList = mutableListOf(itemOne)
            val tempItemLists = mutableListOf<Item>()

            every { viewModel.viewData } returns itemList
            every { viewModel.tempList } returns tempItemLists

            assertThrows<java.lang.IndexOutOfBoundsException> {
                logic.delete(invalidPosition, adapter)
            }
        }
    }


    @Nested
    inner class UndoDeletion {
        /**
         * - Get the last added Item from the ViewModel's temp list.
         * - Re-add to the adapter using the temp position from the ViewModel.
         * - Re-add to the ViewData List.
         * - Remove the last Item from the temp List.
         * - Because the temp List is not empty, pass the temp List to the repo for deletion.
         * - Clear the temp List.
         */
        @Test
        fun undoRecentDeletion() {
            val itemList = mutableListOf<Item>()
            val tempItemList = mutableListOf(itemOne, itemTwo)
            val tempPosition = 0

            every { viewModel.viewData } returns itemList
            every { viewModel.tempList } returns tempItemList
            every { viewModel.tempPosition } returns tempPosition

            logic.undoRecentDeletion(adapter)

            verify { adapter.reAdd(tempPosition, itemTwo) }
            assertThat(itemList).containsExactly(itemTwo)
            verify { repo.deleteItems(tempItemList) }
            assertThat(tempItemList.isEmpty()).isTrue()
        }

        /**
         * - The temp List in the ViewModel is empty.
         * - Exception is thrown
         */
        @Test
        fun `undoRecentDeletion - Empty Temp List`() {
            every { viewModel.tempList } returns emptyList
            every { viewModel.errorMsgInvalidUndo } returns message

            assertThrows<UnsupportedOperationException> {
                logic.undoRecentDeletion(adapter)
            }
        }

        /**
         * - The temp position in the ViewModel is invalid.
         * - Exception is thrown
         */
        @Test
        fun `undoRecentDeletion - Invalid Temp Position`() {
            val invalidPosition = -1
            val tempItemLists = mutableListOf(itemOne)

            every { viewModel.tempList } returns tempItemLists
            every { viewModel.tempPosition } returns invalidPosition
            every { viewModel.errorMsgInvalidUndo } returns message

            assertThrows<UnsupportedOperationException> {
                logic.undoRecentDeletion(adapter)
            }
        }
    }


    @Nested
    inner class DeleteNotificationTimeOut {
        /**
         * - Check if the temp List is empty.
         *   - Will not be in this test.
         * - Pass the temp List [IRepositoryContract.Repository.deleteItems].
         * - Clear the temp List.
         */
        @Test
        fun deletionNotificationTimedOut() {
            val itemList = mutableListOf(itemOne)

            every { viewModel.tempList } returns itemList

            logic.deletionNotificationTimedOut()

            verify { repo.deleteItems(itemList) }
            assertThat(itemList.isEmpty()).isTrue()
        }

        /**
         * - Check if the temp List is empty.
         * - Return without invoking any other method.
         */
        @Test
        fun `deletionNotificationTimedOut - Empty Temp List`() {
            every { viewModel.tempList } returns emptyList

            logic.deletionNotificationTimedOut()

            verify { repo wasNot Called }
        }
    }


    /**
     * - Verify that [CompositeDisposable.clear] is invoked.
     */
    @Test
    fun onDestroy() {
        logic.onDestroy()

        verify { disposable.clear() }
    }
}